### DISCUSSION TITLE
How to determine embedded language at cursor position?

### DISCUSSION THREAD

**Discussion by tinny-blueleo** [nvim-treesitter/nvim-treesitter#6643] [Q&A] [answered]
Hi friends, for embedded languages (e.g. Javascript inside of HTML or Typescript inside of Svelte, etc) I am trying to determine the name of the language at the cursor position.  This is for feeding the language name into other plugins like Luasnip to get snippets.

Currently I have this code:

`
    
local node = ts_utils.get_node_at_cursor(0)

    local lang

    if node ~= nil then
        lang = node:lang()
    else
        lang = vim.bo.filetype
    end

    print('Language at cursor: ' .. lang)
    return lang

`

However I fail getting a parser with this error when trying to get node at the cursor code line:

`
E5108: Error executing lua: ...-linux64/share/nvim/runtime/lua/vim/treesitter/query.lua:259: query: invalid structure at position 3272 for language lua
stack traceback:
        [C]: in function '_ts_parse_query'
        ...-linux64/share/nvim/runtime/lua/vim/treesitter/query.lua:259: in function 'get'
        ...4/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:114: in function 'new'
        ...m/nvim-linux64/share/nvim/runtime/lua/vim/treesitter.lua:61: in function '_create_parser'
        ...m/nvim-linux64/share/nvim/runtime/lua/vim/treesitter.lua:131: in function 'get_parser'
`

Do you know how I can successfully get the language at the cursor or at least get unstuck from this error?  Any help would be greatly appreciated.

**Comment by greg-grigorjevs** [comment, marked as answer]
Here's what worked for me: 
```lua
              local curline = vim.fn.line(".")
              local lang = vim.treesitter.get_parser():language_for_range({curline, 0, curline, 0}):lang()
```

**Reply by tinny-blueleo** [reply]
This works awesome, thank you so much!

**Comment by clason** [comment]
See https://github.com/neovim/neovim/blob/724d1110b1e4699a34f489e9cdb2d25098746499/runtime/lua/vim/_comment.lua#L35 (ignore the part about the commentstring afterwards).

**Comment by sghng** [comment]
I've come a long way to find this discussion. Wouldn't it be nice if we can have a util function to determine the language/filetype under cursor? This would be particularly helpful for anything related to embedded language.
